crates/bevy_render/src/render_resource/buffer_vec.rs

```diff
@@ -511,3 +511,181 @@ where
+    /// size of `T`
+    pub fn new(buffer_usage: BufferUsages, render_device: &RenderDevice) -> Self {
+        let item_size = size_of::<T>();
+
+        let alignment = if cfg!(target_abi = "sim") {
+            // On iOS simulator on silicon macs, metal validation check that the host OS alignment
+            // is respected, but the device reports the correct value for iOS, which is smaller.
+            // Use the larger value.
+            // See https://github.com/gfx-rs/wgpu/issues/7057 - remove if it's not needed anymore.
+            AlignmentValue::new(256)
+        } else if buffer_usage.contains(BufferUsages::UNIFORM) {
+            AlignmentValue::new(render_device.limits().min_uniform_buffer_offset_alignment as u64)
+        } else if buffer_usage.contains(BufferUsages::STORAGE) {
+            AlignmentValue::new(render_device.limits().min_storage_buffer_offset_alignment as u64)
+        } else {
+            unimplemented!(
+                "buffer_usage does not have offset alignment limits. Use RawBufferVec instead."
+            )
+        };
+
+        let aligned_size = alignment.round_up(item_size as u64);
+        let required_padding = aligned_size - item_size as u64;
+        Self {
+            values: Vec::new(),
+            buffer: None,
+            capacity: 0,
+            buffer_usage,
+            label: None,
+            changed: false,
+            aligned_size,
+            required_padding,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Returns a handle to the buffer, if the data has been uploaded.
+    #[inline]
+    pub fn buffer(&self) -> Option<&Buffer> {
+        self.buffer.as_ref()
+    }
+
+    /// Returns the binding for the buffer if the data has been uploaded.
+    #[inline]
+    pub fn binding(&self) -> Option<BindingResource> {
+        Some(BindingResource::Buffer(BufferBinding {
+            buffer: self.buffer.as_ref()?,
+            offset: 0,
+            size: Some(NonZero::new(self.aligned_size)?),
+        }))
+    }
+
+    /// Returns the amount of space that the GPU will use before reallocating.
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Returns the number of items that have been pushed to this buffer.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.values.len() / self.aligned_size as usize
+    }
+
+    /// Returns true if the buffer is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    /// Adds a new value and returns its offset in bytes.
+    ///
+    /// This will convert the value to bytes and add any padding it needs to respect alignment
+    pub fn push(&mut self, value: T) -> usize {
+        let index = self.values.len();
+        let bytes: &[u8] = bytemuck::bytes_of(&value);
+        self.values.extend(bytes);
+        if self.required_padding > 0 {
+            for _ in 0..self.required_padding {
```

Comment:
You could use Vec’s resize to len + required padding.

Reply:
Good idea, done.